moshi/src/main/java/com/squareup/moshi/JsonValueReader.java

```diff
@@ -49,17 +50,31 @@
   /** Sentinel object pushed on {@link #stack} when the reader is closed. */
   private static final Object JSON_READER_CLOSED = new Object();
 
-  private Object[] stack = new Object[32];
+  private Object[] stack;
 
   JsonValueReader(Object root) {
     scopes[stackSize] = JsonScope.NONEMPTY_DOCUMENT;
+    stack = new Object[32];
     stack[stackSize++] = root;
   }
 
+  /** Copy-constructor makes a deep copy for peeking. */
+  JsonValueReader(JsonValueReader copyFrom) {
+    super(copyFrom);
+
+    stack = copyFrom.stack.clone();
+    for (int i = 0; i < stackSize; i++) {
```

Comment:
style nit: use enhanced for loop for the array?

Reply:
it’s slightly different semantics ’cause the loop doesn’t span the physical storage!

Reply:
whoops, yes, ofc.